Optimize the initialization of `List[ndarray]` in `cupy.array`

okapies opened the pull request:
- Implement `cupy.array(<List[cupy.ndarray]>)`
- Improve the performance of `cupy.array(<List[numpy.ndarray]>)`

Fixes #409 and #1084. Ref #1030.

okapies commented:
A performance testing at https://github.com/cupy/cupy/pull/2081/commits/49668ab3e68d71ce0545e7339900589689854aaa (The script is [here](https://gist.github.com/okapies/a6a9d8adbbb29b680f4e78cca44e7221)).

## `--shape="(28, 28)"`
| batch-size | 10 | 100 | 1000 |
|:--|:--|:--|:--|
| numpy -> numpy | 0.153 | 1.246 | 12.679 |
| numpy -> cupy | 1.479 | 3.212 | 19.265 |
| cupy -> cupy | 2.718 | 4.770 | 23.148 |

## `--shape="(3, 224, 224)"`
| batch-size | 1 | 10 | 100 |
|:--|:--|:--|:--|
| numpy -> numpy | 0.454 | 3.913 | 190.949 |
| numpy -> cupy | 2.097 | 9.493 | - |
| cupy -> cupy | 1.504 | 2.775 | 8.908 |

okapies commented:
I run repeated benchmarks for `numpy.array(<List[numpy.ndarray]>)` in a variety of data size. It seems that zero filling of allocated page generates huge overhead when copying large arrays.

![screenshot_2019-03-07 numpy_array_perf](https://user-images.githubusercontent.com/657642/53942519-55640580-40fe-11e9-853c-356fecd5775f.png)

![screenshot_2019-03-07 np_nested_array_125000_numactl log folded gz - speedscope](https://user-images.githubusercontent.com/657642/53943157-d53e9f80-40ff-11e9-8b17-907404c2f226.png)

(Note that you should use `numactl --cpunodebind=0 --membind=0` in a NUMA environment to prevent the unpredictable result.)

okapies commented:
Upon further investigation, dTLB misses and LLC cache misses cause the above performance degradation.

okapies commented:
Specifying `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1` significantly reduces the `dTLB-(load|store)-misses` and improve the overhead for larger array. I'll use it for the benchmark of this PR to measure the baseline performance of my algorithm.

(Note that this configuration leads a fragmentation of the heap memory. Don't use it in production.)

okapies commented:
A performance testing at https://github.com/cupy/cupy/pull/2081/commits/49668ab3e68d71ce0545e7339900589689854aaa (The script is [here](https://gist.github.com/okapies/a6a9d8adbbb29b680f4e78cca44e7221)).

## Parameters
- xp: `numpy` or `cupy`
- shape per item: `(3, 224, 224)`
- batch-size: `1`, `10` or `100`
- number: `10000`
- environment: `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1 numactl --cpunodebind=0 --membind=0`

##  cupy-cuda==6.0.0b3
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `numpy.array(<List[numpy.ndarray]>)` | 0.466 | 3.642 | 133.286 | maybe CPU bootleneck on `memmove` |
| `cupy.array(<List[numpy.ndarray]>)` | 1.201 | 8.308 | 199.298 | 〃 |
| `cupy.array(<List[cupy.ndarray]>)` | n/a | n/a | n/a | - |

## b983599
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | - | - | - | maybe CPU bootleneck on `memmove` |
| `cupy.array(<List[cupy.ndarray]>)` | - | - | - | - |

okapies commented:
## b212555
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | 0.759 | 6.249 | 137.616 | 25-40% faster |

okapies commented:
## Parameters
- xp: `numpy` or `cupy`
- shape per item: `(3, 224, 224)`
- batch-size: `1`, `10` or `100`
- number: `10000`
- environment: `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1 numactl --cpunodebind=0 --membind=0`

## dd5ccbc
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | 0.865 | 6.175 | 136.076 | 25-40% faster than the baseline |
| `cupy.array(<List[cupy.ndarray]>)` | 0.518 | 1.177 | 8.864 | - |

okapies commented:
@niboshi Could you please check it? Thanks.

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+    return a
+
+
+cpdef inline _alloc_pinned_memory(Py_ssize_t nbytes):
```
Isn't it possible to make this function do the transfer as well, not only allocation?
The warning looks out of place.

okapies replied:
Now this is used by both `_send_object_to_gpu` and `_send_numpy_array_list_to_gpu` which have different transfer logics. One idea is that move the warning to the above functions. Do you have any ideas?

okapies replied:
I remembered that it is hard to move to other places because the message takes `e` as an argument.

okapies replied:
Renamed to `_alloc_async_tranfer_buffer` for clarity.

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+        pinned_memory._add_to_watch_list(stream.record(), mem)
+    else:
+        # fallback to numpy array and send it to GPU
+        a = _send_object_to_gpu(arrays, dtype, order, ndmin)
```
Is this fallback meaningful?
Doesn't it cause OOM again in `_send_object_to_gpu`?

okapies replied:
I'll fix to call `a.data.copy_from_host()` directly.

niboshi commented on cupy/core/core.pxd:
```diff
@@ -94,6 +94,14 @@ cdef class Indexer:
     cdef CPointer get_pointer(self)
 
 
+cdef tuple _get_concat_shape(object obj)
```
I think all of the declarations in `.pxd` are not needed.

okapies replied:
Thanks. I'll remove them.

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1708,7 +1710,7 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
             a = src.copy(order=order).astype(dtype, copy=False)
 
-        ndim = a._shape.size()
+        ndim = len(a._shape)
```
Is there any reason of this change? Is `len` faster?

okapies replied:
It is because I saw that the compiler complains `a._shape` doesn't have `size()` method (maybe it fails to infer `_shape` as a tuple). I now checked again and `.size()` works... It is strange. Anyway, I'll fix it.

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+    #    which contains NumPy/CuPy array(s) with same shape and dtype).
+    #    Returns None otherwise.
+    # 2. dtype if the object is an array
+    # 3. a first element in the object
```
How about using "item" instead of "element"?
"Element" more commonly refers to a scalar in arrays.

niboshi replied:
It looks only its type is used.
How about returning the type instead?

niboshi commented on cupy/testing/array.py:
```diff
@@ -128,6 +129,70 @@ def assert_array_list_equal(xlist, ylist, err_msg='', verbose=True):
+        return ctypes.string_at(a.ctypes.data, a.nbytes)
+
+
+def assert_array_exactly_equal(x, y, err_msg='', verbose=True):
```
Does this function compare garbage bytes between elements?
Is there any point to do that? How about comparing strides instead?

okapies commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+    cdef vector.vector[Py_ssize_t] a_shape = (<object>a_cpu).shape
+    cdef ndarray a = ndarray(a_shape, dtype=a_dtype, order=order)
+    if a_cpu.ndim == 0:
+        a.fill(a_cpu[()])
```
Hmm, it was introduced about two years ago in #157. I cannot judge this indexing (`()`) is redundant or not. How do you think?

okapies replied:
As an offline discussion, we found it can be replaced with `a.fill(a_cpu)` because `fill` calls `a_cpu.item()` internally.

okapies commented:
@niboshi Could you please check it again?

niboshi commented on cupy/testing/array.py:
```diff
@@ -128,6 +128,48 @@ def assert_array_list_equal(xlist, ylist, err_msg='', verbose=True):
             verbose=verbose)
 
 
+def assert_array_exactly_equal(x, y, err_msg='', verbose=True):
```
How about adding `strides_check` option in `assert_array_equal`, instead of creating a new function?

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,190 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+    cdef Py_ssize_t nbytes = a.nbytes
+
+    stream = stream_module.get_current_stream()
+    cdef pinned_memory.PinnedMemoryPointer mem = \
```
It's better to use `(` for line continuation instead of `\`. (PEP-8)

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,190 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+    cdef size_t nbytes = itemcount * get_dtype(dtype).itemsize
+
+    stream = stream_module.get_current_stream()
+    cdef pinned_memory.PinnedMemoryPointer mem = \
```
Ditto.

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,190 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
     return a
 
 
+cpdef inline _alloc_async_transfer_buffer(Py_ssize_t nbytes):
```
`cpdef` -> `cdef`?

niboshi commented on cupy/testing/array.py:
```diff
@@ -74,12 +74,14 @@ def assert_array_max_ulp(a, b, maxulp=1, dtype=None):
 
 
-def assert_array_equal(x, y, err_msg='', verbose=True):
+def assert_array_equal(x, y, strides_check=False, err_msg='', verbose=True):
```
The new argument should come last, because the order of arguments is a part of API.

niboshi commented on cupy/testing/array.py:
```diff
@@ -90,6 +92,32 @@ def assert_array_equal(x, y, err_msg='', verbose=True):
+                msg.append(" x: {}".format(x.strides))
+                msg.append(" y: {}".format(y.strides))
+            raise AssertionError('\n'.join(msg))
+        if x.flags.c_contiguous != y.flags.c_contiguous:
```
I think this is redundant. `flags.c_contiguous` and `flags.f_contiguous` are just reduced representations of `strides`.

niboshi commented on cupy/testing/array.py:
```diff
@@ -90,6 +92,32 @@ def assert_array_equal(x, y, err_msg='', verbose=True):
 
+    if strides_check:
+        if x.strides != y.strides:
+            msg = ["Strides are not equal:"]
```
Please use single quotes for string literals.

niboshi commented on cupy/testing/helper.py:
```diff
@@ -337,8 +337,8 @@ def check_func(x, y):
 def numpy_cupy_array_equal(err_msg='', verbose=True, name='xp',
-                           type_check=True, accept_error=False, sp_name=None,
-                           scipy_name=None):
+                           type_check=True, strides_check=False,
```
Ditto. The order of arguments.

niboshi commented on tests/cupy_tests/random_tests/test_generator.py:
```diff
@@ -100,7 +100,8 @@ def _generate_check_repro(self, func, seed):
         testing.assert_array_equal(
             x, y,
-            'Randomly generated arrays with the same seed did not match')
+            err_msg='Randomly generated arrays with the same seed did not '
```
I think this change can be reverted.

niboshi commented:
Jenkins, test this please

okapies commented:
Thanks. I'm checking regressions claimed by Jenkins.

okapies commented:
Jenkins, test this please

okapies commented:
@niboshi I noticed that the new implementation requires `numpy>=1.14.0` (released in Jan 2018) because it depends on third `out` argument of [`numpy.concatenate()`](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.concatenate.html#numpy.concatenate) to prevent memory allocation. Do you think it is OK or not?

okapies commented:
Jenkins, test this please

okapies commented:
## Parameters
- xp: `numpy` or `cupy`
- shape per item: `(3, 224, 224)`
- batch-size: `1`, `10` or `100`
- number: `10000`
- environment: `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1 numactl --cpunodebind=0 --membind=0`

## 756c54a
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | 0.949 | 6.355 | 136.538 | 25-32% faster than the baseline |
| `cupy.array(<List[cupy.ndarray]>)` | 0.550 | 1.184 | 8.862 | - |

okapies commented:
Jenkins, test this please

okapies commented:
Jenkins, test this please.

okapies commented:
@niboshi I added several tests about the handling of scalar value and dtype, and fixed the issues in 956eb2f...5f8af60 (Jenkins failures are the precision problem). Could you please check it and merge if you don't find any other problems?

okapies commented:
Jenkins, test this please.

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
                      dtype, copy, order, subok, ndmin)
+    else:  # obj is sequence, numpy array, scalar or the other type of object
+        shape, elem_type, elem_dtype = _get_concat_shape(obj)
+        if shape is not None and shape[-1] != 0:
```
Doesn't it cause index error if shape is `()`?

okapies replied:
`shape` is always non-empty tuple because a scalar value is *not* concatenatable and `_get_concat_shape()` always return `None` instead of `()`. I added `test_array_from_numpy_scalar` in fe85a30 to check this behavior:

```python
>>> a = np.array(0, dtype='b')
>>> a.ndim
0
>>> a.shape
()
>>> cupy.array(a)
array(0, dtype=int8)
>>> cupy.array([a])
array([0], dtype=int8)
```

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+                a = (_manipulation.concatenate_method(lst, 0)
+                                  .reshape(shape)
+                                  .astype(dtype, order=order, copy=False))
+            else:  # should not be reached here
```
I think it **can** reach here depending on the input.
Please remove this comment.

okapies replied:
`isinstance(elem_type, (numpy.ndarray, cupy.ndarray))` is always true when `shape != None`. If the user sees this exception, it must be a bug of `_get_concat_shape()`.

niboshi replied:
Then, please use `assert` instead of exception.

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+                                  .astype(dtype, order=order, copy=False))
+            else:  # should not be reached here
+                raise ValueError(
+                    'The elements of obj are unsupported type(s)')
```
Wording suggestion:
```suggestion
                    'Elements of obj have unsupported type(s)')
```

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
+            # - scalar or sequence of scalar
+            # - empty sequence or sequence with elements whose shapes or
+            #   dtypes are unmatched
+            # - the other type of object
```
A nitpick but to make it clearer:
```suggestion
            # - other types
```

niboshi commented on cupy/core/core.pyx:
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
     return a
 
 
+cdef bint _is_outable_numpy_concatenate = (
```
How about `_numpy_concatenate_has_out_argument`?

niboshi commented on cupy/testing/array.py:
```diff
@@ -74,12 +74,14 @@ def assert_array_max_ulp(a, b, maxulp=1, dtype=None):
 
 
-def assert_array_equal(x, y, err_msg='', verbose=True):
+def assert_array_equal(x, y, err_msg='', verbose=True, strides_check=False):
```
Could you separate this change in another PR?

okapies replied:
Separated these changes to #2150.

niboshi commented on cupy/testing/helper.py:
```diff
@@ -624,11 +626,14 @@ def _make_all_dtypes(no_float16, no_bool, no_complex):
 
 def for_all_dtypes(name='dtype', no_float16=False, no_bool=False,
-                   no_complex=False):
+                   no_complex=False, no_char_code=True):
```
Ditto. Please separate this change.
We've recently decided to make separate PRs for separate topics as much as possible.

okapies replied:
Separated these changes to #2152.

okapies commented:
@niboshi Hi, could you please check it again?

niboshi commented:
Jenkins, test this please

niboshi commented:
LGTM
I'm sorry for taking time!
